### DISCUSSION TITLE
Menu - using styled-components' forwardedAs with submenu and focus-visible styles

### DISCUSSION THREAD

**Discussion by binhxn** [ariakit/ariakit#1523] [Questions] [answered]
Hello!

I noticed the submenu implementation is different between Ariakit and Reakit. I like using Reakit's example because it provides users access to `useMenuState` and the flexibility to pass any child component to `MenuItem` flexibly. I was hoping I can apply the same approach to Ariakit components, but I'm running into some issues and I'm not sure if it's a `styled-components`-specific issue. 

Here's my [Codesandbox implementation](https://codesandbox.io/s/ariakit-submenu-example-sonxw2?file=/src/App.tsx). I've created a styled-wrapper around Menu components in the `/components` folder.

Issues I'm running into:

1. When attempting to use the `as` prop to reference the Submenu component, the Submenu does not appear. However, when using `forwardedAs`, it works as expected (which is great!), but I'm inheriting styles from both `MenuButton` and `MenuItem`. I tried to use a styled-MenuItem wrapper (`SubmenuButton`) as a workaround but that didn't open the submenu. I'm assuming only the MenuButton can open menus.
2. With the `forwardedAs` prop fixing the Submenu issue... focus-visible styles are appearing when hovering over menu items after initial keyboard navigation. In other words, after using arrow keys to navigate the menu items, hovering the menu items with a mouse seems to activate the focus-visible styles. I'm wondering if it's an issue with `forwardedAs`. In the recording below, notice the focus-visible styles on the **Preferences** item.

https://user-images.githubusercontent.com/8155387/174459974-496be372-878e-4f40-ac23-b33beb90a239.mp4

3. Not styled-components related, but I have two TypeScript errors on line 16 and 45 and was wondering if you had advice on how to approach that 😅 

I'd love to hear your thoughts on this!

**Comment by diegohaz** [comment, marked as answer]
> 1. I'm assuming only the MenuButton can open menus.

That's correct.

> 2\. With the `forwardedAs` prop fixing the Submenu issue... focus-visible styles are appearing when hovering over menu items after initial keyboard navigation. In other words, after using arrow keys to navigate the menu items, hovering the menu items with a mouse seems to activate the focus-visible styles. I'm wondering if it's an issue with `forwardedAs`. In the recording below, notice the focus-visible styles on the **Preferences** item.

That's how `:focus-visible` works. That also happens on the Ariakit example without SC. I recommend using `[data-active-item]` instead of `:focus`, `:hover` and `:focus-visible`.

```diff
- &:hover,
- &:focus {
-   background-color: #f5f5f7;
- }
-
- &:focus-visible {
-   box-shadow: 0 0 0 3px #47b6de inset;
- }
+ &[data-active-item] {
+   background-color: #f5f5f7;
+ }
```

We also have a `data-focus-visible` attribute, but that works the same as the native `:focus-visible` pseudo-class in this case. I'm not sure if we should handle this as a special case. Anyway, I think it's worth opening a separate issue if you have more thoughts.

> 3\. Not styled-components related, but I have two TypeScript errors on line 16 and 45 and was wondering if you had advice on how to approach that 😅

I recommend following the code on https://ariakit.org/examples/menu-nested. Note that it doesn't use `MenuProps` or `MenuButtonProps`. Defining your own prop types will usually give you more control.

Also note that `Submenu` passes the props down to `MenuButton`, not to `Menu`.

```tsx
type SubmenuProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  myCustomProp?: boolean;
};

const Submenu = forwardRef<HTMLButtonElement, SubmenuProps>((props, ref) => {
  ...
});
```

**Reply by binhxn** [reply]
Awesome, this was very helpful. Thank you for your help!
